### PR TITLE
libtorrentRasterbar: fix build, restructure

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change
9c0ebbe7a36d7296e81d97c85cba35aee7bd8334 did not update the patch sha256 hash. (properly?)
Also fails to build anyway when updated.
Thus why I am opting for a temporary unstable version to take the place of 1.1.7

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by worldofpeace** [comment]
Sorry I don't follow what's wrong here?

**Comment by Chiiruno** [comment]
`nix-shell -p nox --run "nox-review wip"` does not finish, but I don't think it's this commit's fault.

**Comment by Chiiruno** [comment]
It didn't build.

**Comment by worldofpeace** [comment]
Hmm, I build it from my updated nixpkgs checkout like `nix-build default.nix -A libtorrentRasterbar`
and it built...

**Comment by Chiiruno** [comment]
Not sure what to say, doesn't work for me. Now it does.

**Comment by worldofpeace** [comment]
Do you have the boost errors output?

**Comment by Chiiruno** [comment]
For the sha256 issue:
```
nix-build . -A libtorrentRasterbar     
warning: Nix search path entry '/home/okina/.nix-defexpr/channels' does not exist, ignoring
these derivations will be built:
  /nix/store/1nh4aakzz3ykxf63gi23jscvk9pzp2sq-64d6b4900448097b0157abb328621dd211e2947d.patch.drv
  /nix/store/i7jfadsgb649zlinxg6rx87wwqa9lynj-libtorrent-rasterbar-1.1.7.drv
building '/nix/store/1nh4aakzz3ykxf63gi23jscvk9pzp2sq-64d6b4900448097b0157abb328621dd211e2947d.patch.drv'...

trying https://github.com/arvidn/libtorrent/commit/64d6b4900448097b0157abb328621dd211e2947d.patch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   902    0   902    0     0   3256      0 --:--:-- --:--:-- --:--:--  3256
fixed-output derivation produced path '/nix/store/rkbvi0pdpqzsfksazas2dk52dn8s1ixh-64d6b4900448097b0157abb328621dd211e2947d.patch' with sha256 hash '1bdv0icqzbg1il60sckcly4y22lkdbkkwdjadwdzxv7cdj586bzd' instead of the expected hash '0d4h0g129rsgm8xikybxypgv6nnya7ap7kskl7q78p4h6y2a0fhc'
cannot build derivation '/nix/store/i7jfadsgb649zlinxg6rx87wwqa9lynj-libtorrent-rasterbar-1.1.7.drv': 1 dependencies couldn't be built
error: build of '/nix/store/i7jfadsgb649zlinxg6rx87wwqa9lynj-libtorrent-rasterbar-1.1.7.drv' failed
```

For the Boost issue:
```
CXX      ../ed25519/src/add_scalar.lo
  CXX      ../ed25519/src/fe.lo
  CXX      ../ed25519/src/ge.lo
  CXX      ../ed25519/src/key_exchange.lo
kademlia/routing_table.cpp: In member function 'libtorrent::dht::routing_table::add_node_status_t libtorrent::dht::routing_table::add_node_impl(libtorrent::dht::node_entry)':
kademlia/routing_table.cpp:716:33: error: 'next' is not a member of 'boost'
  bool const can_split = (boost::next(i) == m_buckets.end()
                                 ^~~~
kademlia/routing_table.cpp:716:33: note: suggested alternative: 'get'
  bool const can_split = (boost::next(i) == m_buckets.end()
                                 ^~~~
                                 get
kademlia/routing_table.cpp:719:40: error: 'prior' is not a member of 'boost'
   && (i == m_buckets.begin() || boost::prior(i)->live_nodes.size() > 1);
                                        ^~~~~
kademlia/routing_table.cpp:907:7: error: in argument to unary !
  if (!can_split)
       ^~~~~~~~~
  CXX      ../ed25519/src/keypair.lo
  CXX      ../ed25519/src/sc.lo
  CXX      ../ed25519/src/seed.lo
  CXX      ../ed25519/src/sha512.lo
make[1]: *** [Makefile:1038: kademlia/routing_table.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/build/libtorrent-libtorrent-1_1_7/src'
make: *** [Makefile:620: all-recursive] Error 1
builder for '/nix/store/0djb1zv8cc3ck2ccggbpvlxv4faq772a-libtorrent-rasterbar-1.1.7.drv' failed with exit code 2
error: build of '/nix/store/0djb1zv8cc3ck2ccggbpvlxv4faq772a-libtorrent-rasterbar-1.1.7.drv' failed
```

**Comment by worldofpeace** [comment]
Weird the patch has a different store path for me  
```
/nix/store/zf7mlvhkghcg5z983cq70rfyn7q3irlx-64d6b4900448097b0157abb328621dd211e2947d.patch
```

**Comment by Chiiruno** [comment]
Should be the same patch, all I did was change the sha256 sum.

**Comment by worldofpeace** [comment]
Perhaps this [patch](https://github.com/arvidn/libtorrent/commit/9cd0ae67e74a507c1b9ff9c057ee97dda38ccb81) is was needed also.
I don't think building from master is the best idea being that it's not production ready.

**Comment by Chiiruno** [comment]
I'll try that out.

**Comment by worldofpeace** [comment]
I can reproduce the error you were getting with master now.

**Comment by Chiiruno** [comment]
Interesting. Must be some local thing.

**Comment by worldofpeace** [comment]
Is there even any benefit to making this build with latest boost and patching it?
I'm reconsidering this https://github.com/NixOS/nixpkgs/pull/42401
It will work with latest boost when they release a new version.

**Comment by Chiiruno** [comment]
Perhaps. I'll include it if you write it.

**Comment by Chiiruno** [comment]
Ah, maybe you want me to include this? https://github.com/arvidn/libtorrent/commit/64d6b4900448097b0157abb328621dd211e2947d

**Comment by Chiiruno** [comment]
Stable (1.1.7) builds now!

**Comment by worldofpeace** [comment]
Hah we're a mess. :1st_place_medal: 

@matthewbauer can you, with your powers, command the @GrahamcOfBorg to do the thing?

**Comment by Chiiruno** [comment]
Do the thing! Do the thing!

**Comment by andir** [comment]
@grahamcofborg build libtorrent-rasterbar

**Comment by andir** [comment]
@GrahamcOfBorg build libtorrentRasterbar

File: pkgs/development/libraries/libtorrent-rasterbar/default.nix [hunk 1]
```diff
@@ -32,12 +38,8 @@ patches = [
     "--with-libiconv=yes"
   ];
 
-  enableParallelBuilding = true;
-
-  doCheck = false; # fails to link
-
   meta = with stdenv.lib; {
-    homepage = http://www.rasterbar.com/products/libtorrent/;
+    homepage = "http://www.rasterbar.com/products/libtorrent";
```
Comment (andir) [inline pkgs/development/libraries/libtorrent-rasterbar/default.nix]: Why was this changed? For me it just redirects infinitely on both versions.

File: pkgs/development/libraries/libtorrent-rasterbar/default.nix [hunk 2]
```diff
@@ -1,26 +1,32 @@
-{ stdenv, fetchurl, fetchpatch, automake, autoconf, boost, openssl, lib, libtool, pkgconfig, zlib, python, libiconv, geoip, ... }:
+{ stdenv, lib, fetchFromGitHub, fetchpatch, pkgconfig, automake, autoconf, zlib
+, boost, openssl, libtool, python, libiconv, geoip }:
 
 stdenv.mkDerivation rec {
   name = "libtorrent-rasterbar-${version}";
   version = "1.1.7";
+  formattedVersion = lib.replaceChars ["."] ["_"] version;
```
Comment (andir) [inline pkgs/development/libraries/libtorrent-rasterbar/default.nix]: I am (almost always) a fan of having less (garbage/temporary) attributes inside the actual `mkDerivation` attrset. While we are already touching the expression in some parts. How about you change it to something like:

```nix
{…}:
let # the `let` expression is always recursive per default
  version = "1.1.7";
  formattedVersion = lib.replaceChars ["."] ["_"] version;
in
stdenv.mkDerivation { # note the missing `rec` here
  name = "libtorrent-rasterbar-${version}";
  src = fetchFromGitHub { … };
  …
}
```

**Review by andir — CHANGES_REQUESTED** [review]
Thank you for working on this!

There are a few nitpicks from my side:

- Please split the build fix and the change in structure (formattedVersion moved, fetchFromGitHub, …) into separated commits
- Change the commit format so it does follow our guidelines. The messages (for this package) should start with `libtorrentRasterbar` since that is the name within `all-packages.nix`.

File: pkgs/development/libraries/libtorrent-rasterbar/default.nix (hunk 1 above)
Comment (worldofpeace) [reply pkgs/development/libraries/libtorrent-rasterbar/default.nix]: It seems this url was dead before this change was made

File: pkgs/development/libraries/libtorrent-rasterbar/default.nix [hunk 3]
```diff
@@ -1,26 +1,32 @@
...
 
-  src =
-    let formattedVersion = lib.replaceChars ["."] ["_"] version;
-    in fetchurl {
-      url = "https://github.com/arvidn/libtorrent/archive/libtorrent-${formattedVersion}.tar.gz";
-      sha256 = "0vbw7wcw8x9787rq5fwaibpvvspm3237l8ahbf20gjpzxhn4yfwc";
-    };
+  src = fetchFromGitHub {
```
Comment (worldofpeace) [inline pkgs/development/libraries/libtorrent-rasterbar/default.nix]: Thanks for changing this to `fetchFromGitHub`
It was very ugly before.

**Comment by Chiiruno** [comment]
>Please split the build fix and the change in structure (formattedVersion moved, fetchFromGitHub, …) into separated commits

If you really really have to have that, I'll do it, but as far as I can see, this seems unecessary, especially since it's in the same PR.

**Review by worldofpeace — APPROVED** [review]

**Comment by Chiiruno** [comment]
Also couldn't we change platforms to `all` instead of unix since this clearly builds on darwin?

**Comment by worldofpeace** [comment]
> Also couldn't we change platforms to all instead of unix since this clearly builds on darwin?

platforms.unix does include darwin. 
> if you think it's worth it, I can change it to the github or if it has one, a related webpage.

It's good to have working webpages in the meta.

**Review by worldofpeace — APPROVED** [review]

**Comment by Chiiruno** [comment]
>platforms.unix does include darwin.

I'm aware, it's based off of BSD after all.
I was just wondering since isn't UNIX-based all NixOS supports right now?

**Review by worldofpeace — APPROVED** [review]

**Comment by andir** [comment]
@grahamcofborg build libtorrentRasterbar

**Comment by worldofpeace** [comment]
This usually happens when you run out of file descriptors or the borg reached it's limit I think.
Should be unrelated if it actually builds or not.

@andir Could you run the darwin build again?

**Comment by Chiiruno** [comment]
Oh, that must be why my `meguca` package failed too, it failed at a weird place on one build.

**Comment by andir** [comment]
@GrahamcOfBorg build libtorrentRasterbar

**Review by andir — APPROVED** [review]

**Comment by worldofpeace** [comment]
@Chiiruno Yay! :hand:
